0015-authority.md

```diff
@@ -0,0 +1,478 @@
+    {
+        // Owner client gives inputs into network variable.
+        if (IsOwner)
+        {
+            m_MyInput.Value = Input.GetKeyDown(KeyCode.Space);
+        }
+
+        // The authority which could either be the owner or the server based on the authority mode uses the input to move.
+        if(HasAuthority)
+        {
+            if(m_MyInput.Value)
+            {
+                m_Rigidbody.AddForce(transform.up * m_Thrust);
+            }
+        }
+    }
+}
+```
+*The authority system allows NetworkVariables to be flexibly used as inputs.*
+
+## Changing Authority
+
+The `Authority` mode of a `NetworkObject` can be changed, on the server only, by calling `networkObject.ChangeAuthority(Authority.Server);`.
+
+This means which peer has authority also changes if the ownership of an object changes when in `owner` authority mode. This means the `HasAuthority` property should never be cached but always accessed directly.
+
+## OnGainedAuthority / OnLostAuthority
+
+`OnGainedAuthority` and `OnLostAuthority` functions can be overridden to get informed about changes to the `HasAuthority` status. `OnGainedAuthority` gets called when `HasAuthority` switches from false to true. `OnLostAuthority` gets called when it switches to false from true
+```csharp
+public override void OnGainedAuthority()
+{
+    Debug.Log("Authority Gained :)");
+}
+
+public override void OnLostAuthority()
+{
+    Debug.Log("Authority Lost :(");
+}
+```
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Why is Authority a mode?
+Why is authority not something like ownership which you give to a client. The reason for that is because authority is a relation between multiple peers. In an authoritative model a set of peers pass inputs to the authoritative peer for processing.
+
+Because we are a star topology we have some technical limitations. For instance we could not have one client generate the inputs but pass it to another client which holds authority (because this would break the encapsulation principle of time explained in the NetworkTime RFC).
+
+Because of the above we will only support authority models where the owner of an object collects inputs and sends them to the authoritative peer which has to be either the owner or the server.
+
+Accepting inputs from multiple peers is something we will not support with the current proposed design. The authority mode system allows for further extension to support those modes.
+
+## Ownership vs Authority
+This RFC will introduce some rules and responsibilities for each of the two systems.
+
+IsOwner:
+- Only one peer can own a NetworkObject.
+- The `owner` of a NetworkObject can use `ServerRpcs` on its `NetworkBehaviour` by default.
+- The `owner` is responsible for collecting inputs and redirecting them to the authority.
+- The `owner` is encouraged write to `NetworkVariables` with readpermission = authority but not to other NetworkVariables.
```

Comment:
you mean "writepermission = authority", not read permission?

Reply:
I mean `readpermission = authority` as it is there. The `owner` can use NetworkVariables with `readpermission = authority` to send inputs to the authority. That was what I was trying to say here. For this to work the writepermission would have to be set to `owner` of course.

Reply:
you're talking about "read" and "sending inputs". To me those two are in opposition. reading shouldn't imply any changes.

Reply:
Let me try to reformulate this. Does this make more sense?
> If the `owner` wants to pass inputs to the authority this can be done by writing to `NetworkVariables` with WritePermission = owner and ReadPermission = Authority.